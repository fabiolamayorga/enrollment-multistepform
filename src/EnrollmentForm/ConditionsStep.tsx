import React, { useState, useEffect } from 'react';
import { Form, Header, Grid, Container } from 'semantic-ui-react'
import { Field } from 'formik';

import data from '../data/data.json';
import { IConditions, ConditionProps } from '../interfaces/interfaces';


const ConditionsStep: React.FC<ConditionProps> = ({ values }) => {
    const selectedTypeConditions: string[] = values.conditions;
    const initialArray: IConditions[] = [];
    const [filters, setFilters] = useState(initialArray);
    const conditions: IConditions[] = data;

    const uniqueObjects: string[] = [...new Set(conditions.map(x => x.type))];

    useEffect(() => {
        let filteredArray: IConditions[] = [];

        if (selectedTypeConditions) {
            filteredArray = conditions.filter(
                (item) =>
                    selectedTypeConditions.includes(item.type)
            );
            setFilters(filteredArray);
        };
    }, [selectedTypeConditions, conditions]);


    return (
        <Container>
            <Header as='h2'>Conditions</Header>
            <p>Please select your medical conditions:</p>
            <Grid.Row>
                <Form.Group key="conditions">
                    {
                        uniqueObjects.map((type, index) => {
                            return (
                                <Form.Field key={`checkbox-${index}`} inline>
                                    <Field type="checkbox" name="conditions" value={type} />
                                    <label>{type}</label>
                                </Form.Field>

                            )
                        })

                    }
                </Form.Group>
            </Grid.Row>
            <Grid.Row>
                <Form.Group key="conditions" >
                    {
                        <div>
                            {filters.map((item, index) => {
                                return (
                                    <Form.Field key={`subCheckbox-${index}`} inline>
                                        <Field type="checkbox" name="personalConditions" value={item.condition} />
                                        <label>{item.condition}</label>
                                    </Form.Field>
                                )
                            })}
                        </div>
                    }

                </Form.Group>
            </Grid.Row>

        </Container >
    )
}

export default ConditionsStep;