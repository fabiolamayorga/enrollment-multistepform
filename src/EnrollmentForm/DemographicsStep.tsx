import React from 'react';
import { Form, Header, Container } from 'semantic-ui-react';
import { Field } from 'formik';


const DemographicStep: React.FC = () => {
    return (
        <Container>
            <Header as='h2'>Personal Information</Header>
            <Form.Field key="first-name">
                <label>First Name</label>
                <Field placeholder='First Name' name="firstName" id="firstName"/>
            </Form.Field>
            <Form.Field key="last-name">
                <label>Last Name</label>
                <Field placeholder='Last Name' name="lastName" id="lastName"/>
            </Form.Field>
            <Form.Field key="email">
                <label>Email</label>
                <Field placeholder='Email' name="email" id="email"/>
            </Form.Field>
            <Form.Field key="address">
                <label>Address</label>
                <Field placeholder='Address' name="address" id="address"/>
            </Form.Field>
            <Form.Field key="birthdate">
                <label>Birthdate</label>
                <Field placeholder='Birthdate' type="date" name="birthdate"  id="birthdate"/>
            </Form.Field>

        </Container>
    )
}

export default DemographicStep;