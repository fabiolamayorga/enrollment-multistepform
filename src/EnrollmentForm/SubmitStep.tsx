import React from 'react';
import { Card, Container, Divider, Header, Form } from 'semantic-ui-react'
import { ConditionProps } from '../interfaces/interfaces';
import { Field } from 'formik';

const SubmitStep: React.FC = () => {
    return (
        <Container>
            <Header>Submit Step</Header>
            <Divider />
            <Card fluid>
                <Card.Content>
                    <Card.Header content="Terms" />
                    <Card.Description content='Nullam quis risus eget urna mollis ornare vel eu leo. Aenean lacinia bibendum nulla sed consectetur. Integer posuere erat a ante venenatis dapibus posuere velit aliquet. Etiam porta sem malesuada magna mollis euismod. Maecenas sed diam eget risus varius blandit sit amet non magna.' />
                    <Card.Description content='Etiam porta sem malesuada magna mollis euismod. Nullam quis risus eget urna mollis ornare vel eu leo. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Vestibulum id ligula porta felis euismod semper. Donec ullamcorper nulla non metus auctor fringilla.' />

                </Card.Content>
            </Card>
            <Form.Field inline>
                <Field type="checkbox" name="acceptedTerms" value="yes" />
                <label>Agree Terms and Conditions</label>
            </Form.Field>
        </Container>
    )
}

export default SubmitStep;
