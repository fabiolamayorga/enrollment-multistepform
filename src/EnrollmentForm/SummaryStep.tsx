import React from "react";
import { Container, Header, Table } from "semantic-ui-react";
import { ConditionProps } from "../interfaces/interfaces";

const SummaryStep: React.FC<ConditionProps> = ({ values, errors }) => {
  return (
    <Container>
      <Header>Questionary Summary</Header>
      <Table>
        <Table.Header>
          <Table.Row>
            <Table.HeaderCell>Field</Table.HeaderCell>
            <Table.HeaderCell>Response</Table.HeaderCell>
            <Table.HeaderCell>Error</Table.HeaderCell>
          </Table.Row>
        </Table.Header>
        <Table.Body>
          <Table.Row>
            <Table.Cell>First Name</Table.Cell>
            <Table.Cell>{values.firstName}</Table.Cell>
            <Table.Cell>{errors?.firstName}</Table.Cell>
          </Table.Row>
          <Table.Row>
            <Table.Cell>Last Name</Table.Cell>
            <Table.Cell>{values.lastName}</Table.Cell>
            <Table.Cell>{errors?.lastName}</Table.Cell>
          </Table.Row>
          <Table.Row>
            <Table.Cell>Address</Table.Cell>
            <Table.Cell>{values.address}</Table.Cell>
            <Table.Cell>{errors?.address}</Table.Cell>
          </Table.Row>
          <Table.Row>
            <Table.Cell>Birthdate</Table.Cell>
            <Table.Cell>{values.birthdate}</Table.Cell>
            <Table.Cell>{errors?.birthdate}</Table.Cell>
          </Table.Row>
          <Table.Row>
            <Table.Cell>Medical conditions</Table.Cell>
            <Table.Cell>{values.conditions}</Table.Cell>
            <Table.Cell>{errors?.conditions}</Table.Cell>
          </Table.Row>
          <Table.Row>
            <Table.Cell>Personal conditions</Table.Cell>
            <Table.Cell>{values.personalConditions}</Table.Cell>
            <Table.Cell>{errors?.personalConditions}</Table.Cell>
          </Table.Row>
          <Table.Row>
            <Table.Cell>Do you smoke any tobacco products?</Table.Cell>
            <Table.Cell>{values.smoke}</Table.Cell>
            <Table.Cell>{errors?.smoke}</Table.Cell>
          </Table.Row>
          <Table.Row>
            <Table.Cell>Do you drink alcohol?</Table.Cell>
            <Table.Cell>{values.alcohol}</Table.Cell>
            <Table.Cell>{errors?.alcohol}</Table.Cell>
          </Table.Row>
          <Table.Row>
            <Table.Cell>Have you regularly used illicit drugs?</Table.Cell>
            <Table.Cell>{values.drugs}</Table.Cell>
            <Table.Cell>{errors?.drugs}</Table.Cell>
          </Table.Row>
        </Table.Body>
      </Table>
    </Container>
  );
};

export default SummaryStep;
