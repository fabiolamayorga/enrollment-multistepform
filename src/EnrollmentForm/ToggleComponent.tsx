import React, { useState } from 'react';
import { Form, Container, Grid } from 'semantic-ui-react'
import { Field } from 'formik';

import { ToogleProps } from '../interfaces/interfaces';


const ToggleComponent: React.FC<ToogleProps> = ({ question, hiddenQuestion, hiddenQuestionName, name, values }) => {

    const [showHiddenQuestion, setShowHiddenQuestion] = useState(values[name] === 'yes');

    const handleChange = (e: any): void => {
        if (e.target.value === "yes") {
            setShowHiddenQuestion(true);
        } else {
            setShowHiddenQuestion(false);
        }
    }
    return (
        <Container>
            <Grid.Row><label>{question}</label></Grid.Row>
            <Grid.Row>
                <Form.Group onChange={(e: any) => handleChange(e)}>
                    <Form.Field>
                        <label>Yes</label>
                        <Field value="yes"
                            control='input'
                            type='radio'
                            name={name}
                        />
                    </Form.Field>
                    <Form.Field>
                        <label>No</label>
                        <Field value="no"
                            control='input'
                            type='radio'
                            name={name}
                        />
                    </Form.Field>
                </Form.Group>
            </Grid.Row>

            {showHiddenQuestion &&
                <Form.Group>
                    <Form.Field key="first-name">
                        <label>{hiddenQuestion}</label>
                        <Field name={hiddenQuestionName} id={hiddenQuestionName} />
                    </Form.Field>
                </Form.Group>
            }
        </Container>
    )
}

export default ToggleComponent;