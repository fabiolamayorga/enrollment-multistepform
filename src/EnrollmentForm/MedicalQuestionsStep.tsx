import React from 'react';
import { Header, Container, } from 'semantic-ui-react'
import { ConditionProps } from '../interfaces/interfaces';
import ToggleComponent from './ToggleComponent';

const MedicalQuestionsStep: React.FC<ConditionProps> = ({values}) => {
    return (
        <Container>
            <Header as='h2'>Personal Information</Header>

            <ToggleComponent
                question={"Do you smoke any tobacco products?"}
                hiddenQuestion={"How much and how often?"}
                hiddenQuestionName={"howOftenSmoke"}
                name="smoke"
                values={values}
            />

            <ToggleComponent
                question={"Do you drink alcohol?"}
                hiddenQuestion={"How much and how often?"}
                hiddenQuestionName={"howOftenAlcohol"}
                name="alcohol"
                values={values}

            />

            <ToggleComponent
                question={"Have you regularly used illicit drugs?"}
                hiddenQuestion={"How much and how often?"}
                hiddenQuestionName={"howOftenDrugs"}
                name="drugs"
                values={values}

            />
        </Container>
    )
}

export default MedicalQuestionsStep;