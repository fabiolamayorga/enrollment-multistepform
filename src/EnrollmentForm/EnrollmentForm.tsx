import React, { useState } from "react";

import { Button, Form } from "formik-semantic-ui";

import DemographicStep from "./DemographicsStep";
import ConditionStep from "./ConditionsStep";
import MedicalQuestionsStep from "./MedicalQuestionsStep";
import SummaryStep from "./SummaryStep";
import SubmitStep from "./SubmitStep";

import { Errors, IFormConditions } from "../interfaces/interfaces";

import * as Yup from "yup";
import { Divider, Grid } from "semantic-ui-react";

const EnrollmentForm = () => {
  const initialValues: IFormConditions = {
    firstName: "",
    lastName: "",
    email: "",
    address: "",
    birthdate: "",
    conditions: [],
    personalConditions: [],
    smoke: "",
    alcohol: "",
    drugs: "",
    howOftenSmoke: "",
    howOftenAlcohol: "",
    howOftenDrugs: "",
    acceptedTerms: "",
  };

  const validationFormConditionsSchema = Yup.object({
    firstName: Yup.string()
      .max(15, "Must be 15 characters or less")
      .required("Required"),
    lastName: Yup.string()
      .max(20, "Must be 20 characters or less")
      .required("Required"),
    email: Yup.string().email("Invalid email address").required("Required"),
    address: Yup.string()
      .min(5, "Must be 5 characters or more")
      .required("Required"),
    conditions: Yup.array().min(1, "Must be select 1 or more"),
    personalConditions: Yup.array().min(1, "Must be select 1 or more"),
    birthdate: Yup.string().required("Required"),
    smoke: Yup.string().required("Required"),
    howOftenSmoke: Yup.string().when("smoke", {
      is: "yes",
      then: Yup.string().required("Must enter value"),
    }),
    alcohol: Yup.string().required("Required"),
    howOftenAlcohol: Yup.string().when("alcohol", {
      is: "yes",
      then: Yup.string().required("Must enter value"),
    }),
    drugs: Yup.string().required("Required"),
    howOftenDrugs: Yup.string().when("drugs", {
      is: "yes",
      then: Yup.string().required("Must enter value"),
    }),
  });

  const [step, setStep] = useState(0);

  const renderSwitch = (
    errors: Errors,
    values: IFormConditions,
  ) => {
    switch (step) {
      case 0:
        return <DemographicStep />;
      case 1:
        return <ConditionStep values={values} />;
      case 2:
        return <MedicalQuestionsStep values={values} />;
      case 3:
        return <SummaryStep values={values} errors={errors} />;
      case 4:
        return <SubmitStep />;
    }
  };

  const onClickNext = (e: any): void => {
    e.preventDefault();
    setStep(step + 1);
  };

  const onClickPreview = (e: any): void => {
    e.preventDefault();
    setStep(step - 1);
  };

  const onSubmitForm = (values: IFormConditions, formikApi: any): void => {
    console.log(values, formikApi);
    formikApi.setSubmitting(false)
  };

  const validateBack = (): boolean => {
    return step < 1;
  };

  const validateNext = (errors: Errors): boolean => {
    if (step === 3) {
      return Object.keys(errors).length !== 0;
    }

    return step === 4;
  };

  const validateTerms = (values: IFormConditions): boolean => {
    return values?.acceptedTerms[0] !== 'yes'
  };



  return (
    <Form
      initialValues={initialValues}
      validateOnMount
      validationSchema={validationFormConditionsSchema}
      onSubmit={onSubmitForm}
    >
      {({ errors, values }) => (
        <Form.Children>
          {renderSwitch(errors, values)}
          <Divider />
          <Grid.Row>
            <Button disabled={validateBack()} onClick={onClickPreview}>
              Preview
            </Button>

            {step !== 4 && (
              <Button onClick={onClickNext} disabled={validateNext(errors)}>
                Next
              </Button>
            )}
            {step === 4 && <Button.Submit disabled={validateTerms(values)}>Submit</Button.Submit>}
          </Grid.Row>
        </Form.Children>
      )}
    </Form>
  );
};

export default EnrollmentForm;
