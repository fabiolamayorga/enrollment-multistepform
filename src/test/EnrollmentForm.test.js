import renderer from 'react-test-renderer';

import EnrollmentForm from '../EnrollmentForm/EnrollmentForm';

test('Test Links', () => {
    const component = renderer.create(
      <EnrollmentForm />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  
    // manually trigger the callback
    tree.props.onClickNext;
    //re-rendering
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  
    // manually trigger the callback
    tree.props.onClickPreview;
    // re-rendering
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
