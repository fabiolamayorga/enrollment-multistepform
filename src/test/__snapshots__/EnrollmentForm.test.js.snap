// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Links 1`] = `
<form
  className="ui form"
  onSubmit={[Function]}
>
  <div
    className="ui container"
  >
    <h2
      className="ui header"
    >
      Personal Information
    </h2>
    <div
      className="field"
    >
      <label>
        First Name
      </label>
      <input
        id="firstName"
        name="firstName"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="First Name"
        value=""
      />
    </div>
    <div
      className="field"
    >
      <label>
        Last Name
      </label>
      <input
        id="lastName"
        name="lastName"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Last Name"
        value=""
      />
    </div>
    <div
      className="field"
    >
      <label>
        Email
      </label>
      <input
        id="email"
        name="email"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Email"
        value=""
      />
    </div>
    <div
      className="field"
    >
      <label>
        Address
      </label>
      <input
        id="address"
        name="address"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Address"
        value=""
      />
    </div>
    <div
      className="field"
    >
      <label>
        Birthdate
      </label>
      <input
        id="birthdate"
        name="birthdate"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Birthdate"
        type="date"
        value=""
      />
    </div>
  </div>
  <div
    className="ui divider"
  />
  <div
    className="row"
  >
    <button
      className="ui disabled button"
      disabled={true}
      onClick={[Function]}
      tabIndex={-1}
      type="button"
    >
      Preview
    </button>
    <button
      className="ui button"
      onClick={[Function]}
      type="button"
    >
      Next
    </button>
  </div>
</form>
`;

exports[`Test Links 2`] = `
<form
  className="ui form"
  onSubmit={[Function]}
>
  <div
    className="ui container"
  >
    <h2
      className="ui header"
    >
      Personal Information
    </h2>
    <div
      className="field"
    >
      <label>
        First Name
      </label>
      <input
        id="firstName"
        name="firstName"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="First Name"
        value=""
      />
    </div>
    <div
      className="field"
    >
      <label>
        Last Name
      </label>
      <input
        id="lastName"
        name="lastName"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Last Name"
        value=""
      />
    </div>
    <div
      className="field"
    >
      <label>
        Email
      </label>
      <input
        id="email"
        name="email"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Email"
        value=""
      />
    </div>
    <div
      className="field"
    >
      <label>
        Address
      </label>
      <input
        id="address"
        name="address"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Address"
        value=""
      />
    </div>
    <div
      className="field"
    >
      <label>
        Birthdate
      </label>
      <input
        id="birthdate"
        name="birthdate"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Birthdate"
        type="date"
        value=""
      />
    </div>
  </div>
  <div
    className="ui divider"
  />
  <div
    className="row"
  >
    <button
      className="ui disabled button"
      disabled={true}
      onClick={[Function]}
      tabIndex={-1}
      type="button"
    >
      Preview
    </button>
    <button
      className="ui button"
      onClick={[Function]}
      type="button"
    >
      Next
    </button>
  </div>
</form>
`;

exports[`Test Links 3`] = `
<form
  className="ui form"
  onSubmit={[Function]}
>
  <div
    className="ui container"
  >
    <h2
      className="ui header"
    >
      Personal Information
    </h2>
    <div
      className="field"
    >
      <label>
        First Name
      </label>
      <input
        id="firstName"
        name="firstName"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="First Name"
        value=""
      />
    </div>
    <div
      className="field"
    >
      <label>
        Last Name
      </label>
      <input
        id="lastName"
        name="lastName"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Last Name"
        value=""
      />
    </div>
    <div
      className="field"
    >
      <label>
        Email
      </label>
      <input
        id="email"
        name="email"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Email"
        value=""
      />
    </div>
    <div
      className="field"
    >
      <label>
        Address
      </label>
      <input
        id="address"
        name="address"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Address"
        value=""
      />
    </div>
    <div
      className="field"
    >
      <label>
        Birthdate
      </label>
      <input
        id="birthdate"
        name="birthdate"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Birthdate"
        type="date"
        value=""
      />
    </div>
  </div>
  <div
    className="ui divider"
  />
  <div
    className="row"
  >
    <button
      className="ui disabled button"
      disabled={true}
      onClick={[Function]}
      tabIndex={-1}
      type="button"
    >
      Preview
    </button>
    <button
      className="ui button"
      onClick={[Function]}
      type="button"
    >
      Next
    </button>
  </div>
</form>
`;
